Not starting up due to memory violation or nil pointer exception

**Describe the bug**
After a restart of the k3s cluster, I'm unable to start up the `argocd-image-update` due to a memory or nil pointer issue.
Before it was running for month without any issue.

**To Reproduce**
Steps to reproduce the behavior:

1. Deploy default `Application` with no additional settings in the same namespace as you are running your ArgoCD instance
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: argocd-image-updater
  namespace: argocd
  finalizers:
    - resources-finalizer.argocd.argoproj.io
spec:
  project: default
  source:
    repoURL: https://github.com/argoproj-labs/argocd-image-updater.git
    path: manifests
    directory:
      include: install.yaml
    targetRevision: HEAD
  destination:
    server: https://kubernetes.default.svc
    namespace: argocd
```

**Expected behavior**
A clear and concise description of what you expected to happen.

- Clean startup without memory violation or nil pointer exception

**Additional context**
Add any other context about the problem here.

- It's a single node k3s cluster which has been restarted due to an increase (Proxmox VM) of
  - local LVM volume (`/`)
  - add more memory
- No other workloads or system service are affected

**Version**
Please tell us about the version you encountered the issue with

- Image: `quay.io/argoprojlabs/argocd-image-updater:latest`
- ArgoCD: `quay.io/argoproj/argocd:v2.13.3`

**Logs**
```txt
time="2025-01-26T09:44:10Z" level=info msg="argocd-image-updater v99.9.9+980eff5 starting [loglevel:INFO, interval:2m0s, healthport:8080]"
time="2025-01-26T09:44:10Z" level=warning msg="commit message template at /app/config/commit.template does not exist, using default"
time="2025-01-26T09:44:10Z" level=warning msg="Registry configuration at /app/config/registries.conf could not be read: stat /app/config/registries.conf: no such file or directory -- using default configuration"
time="2025-01-26T09:44:10Z" level=info msg="ArgoCD configuration: [apiKind=kubernetes, server=argocd-server.argocd, auth_token=false, insecure=false, grpc_web=false, plaintext=false]"
time="2025-01-26T09:44:10Z" level=info msg="Starting health probe server TCP port=8080"
time="2025-01-26T09:44:10Z" level=info msg="Starting metrics server on TCP port=8081"
time="2025-01-26T09:44:10Z" level=info msg="Warming up image cache"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x22f71b4]

goroutine 1 [running]:
github.com/argoproj-labs/argocd-image-updater/pkg/argocd.(*k8sClient).ListApplications(0x300000002?, {0x0, 0x0})
    /src/argocd-image-updater/pkg/argocd/argocd.go:64 +0x34
main.runImageUpdater(0xc00072cdc0, 0x1)
    /src/argocd-image-updater/cmd/run.go:270 +0x159
main.warmupImageCache(0xc00072cdc0)
    /src/argocd-image-updater/cmd/run.go:365 +0x3e
main.newRunCommand.func1(0xc00011f000?, {0x2a03427?, 0x4?, 0x2a0342b?})
    /src/argocd-image-updater/cmd/run.go:153 +0xd73
github.com/spf13/cobra.(*Command).execute(0xc000227808, {0x47c0ac0, 0x0, 0x0})
    /go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:985 +0xaaa
github.com/spf13/cobra.(*Command).ExecuteC(0xc000227508)
    /go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1117 +0x3ff
github.com/spf13/cobra.(*Command).Execute(...)
    /go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1041
main.newRootCommand()
    /src/argocd-image-updater/cmd/main.go:65 +0x1b4
main.main()
    /src/argocd-image-updater/cmd/main.go:82 +0x4e
stream closed EOF for argocd/argocd-image-updater-bccb9d7d7-72bpg (argocd-image-updater)
```

--- a/.readthedocs.yml
+++ b/.readthedocs.yml
@@ -6,6 +6,7 @@ build:
     python: "3.12"
 mkdocs:
   fail_on_warning: false
+  configuration: mkdocs.yml
 python:
   install:
     - requirements: docs/requirements.txt

--- a/cmd/util.go
+++ b/cmd/util.go
@@ -7,6 +7,7 @@ import (
 	"time"
 
 	"github.com/argoproj-labs/argocd-image-updater/pkg/kube"
+
 	"github.com/argoproj/argo-cd/v2/pkg/client/clientset/versioned"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/tools/clientcmd"

--- a/cmd/util.go
+++ b/cmd/util.go
@@ -3,12 +3,13 @@ package main
 import (
 	"context"
 	"fmt"
-	"path/filepath"
+	"os"
 	"time"
 
 	"github.com/argoproj-labs/argocd-image-updater/pkg/kube"
-	registryKube "github.com/argoproj-labs/argocd-image-updater/registry-scanner/pkg/kube"
-	"github.com/argoproj-labs/argocd-image-updater/registry-scanner/pkg/log"
+	"github.com/argoproj/argo-cd/v2/pkg/client/clientset/versioned"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
 )
 
 func getPrintableInterval(interval time.Duration) string {
@@ -28,30 +29,37 @@ func getPrintableHealthPort(port int) string {
 }
 
 func getKubeConfig(ctx context.Context, namespace string, kubeConfig string) (*kube.ImageUpdaterKubernetesClient, error) {
-	var fullKubeConfigPath string
 	var kubeClient *kube.ImageUpdaterKubernetesClient
-	var err error
 
-	if kubeConfig != "" {
-		fullKubeConfigPath, err = filepath.Abs(kubeConfig)
-		if err != nil {
-			return nil, fmt.Errorf("cannot expand path %s: %v", kubeConfig, err)
-		}
-	}
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	loadingRules.DefaultClientConfig = &clientcmd.DefaultClientConfig
+	loadingRules.ExplicitPath = kubeConfig
+	overrides := clientcmd.ConfigOverrides{}
+	clientConfig := clientcmd.NewInteractiveDeferredLoadingClientConfig(loadingRules, &overrides, os.Stdin)
 
-	if fullKubeConfigPath != "" {
-		log.Debugf("Creating Kubernetes client from %s", fullKubeConfigPath)
-	} else {
-		log.Debugf("Creating in-cluster Kubernetes client")
-	}
-
-	kubernetesClient, err := registryKube.NewKubernetesClientFromConfig(ctx, namespace, fullKubeConfigPath)
+	config, err := clientConfig.ClientConfig()
 	if err != nil {
 		return nil, err
 	}
-	kubeClient = &kube.ImageUpdaterKubernetesClient{
-		KubeClient: kubernetesClient,
+
+	if namespace == "" {
+		namespace, _, err = clientConfig.Namespace()
+		if err != nil {
+			return nil, err
+		}
 	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	applicationsClientset, err := versioned.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	kubeClient = kube.NewKubernetesClient(ctx, clientset, applicationsClientset, namespace)
 
 	return kubeClient, nil
 }

--- a/cmd/util.go
+++ b/cmd/util.go
@@ -3,14 +3,9 @@ package main
 import (
 	"context"
 	"fmt"
-	"os"
 	"time"
 
 	"github.com/argoproj-labs/argocd-image-updater/pkg/kube"
-
-	"github.com/argoproj/argo-cd/v2/pkg/client/clientset/versioned"
-	"k8s.io/client-go/kubernetes"
-	"k8s.io/client-go/tools/clientcmd"
 )
 
 func getPrintableInterval(interval time.Duration) string {
@@ -30,37 +25,11 @@ func getPrintableHealthPort(port int) string {
 }
 
 func getKubeConfig(ctx context.Context, namespace string, kubeConfig string) (*kube.ImageUpdaterKubernetesClient, error) {
-	var kubeClient *kube.ImageUpdaterKubernetesClient
 
-	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
-	loadingRules.DefaultClientConfig = &clientcmd.DefaultClientConfig
-	loadingRules.ExplicitPath = kubeConfig
-	overrides := clientcmd.ConfigOverrides{}
-	clientConfig := clientcmd.NewInteractiveDeferredLoadingClientConfig(loadingRules, &overrides, os.Stdin)
-
-	config, err := clientConfig.ClientConfig()
+	kubeClient, err := kube.NewKubernetesClient(ctx, kubeConfig, namespace)
 	if err != nil {
 		return nil, err
 	}
-
-	if namespace == "" {
-		namespace, _, err = clientConfig.Namespace()
-		if err != nil {
-			return nil, err
-		}
-	}
-
-	clientset, err := kubernetes.NewForConfig(config)
-	if err != nil {
-		return nil, err
-	}
-
-	applicationsClientset, err := versioned.NewForConfig(config)
-	if err != nil {
-		return nil, err
-	}
-
-	kubeClient = kube.NewKubernetesClient(ctx, clientset, applicationsClientset, namespace)
 
 	return kubeClient, nil
 }

--- a/pkg/kube/kubernetes.go
+++ b/pkg/kube/kubernetes.go
@@ -5,6 +5,7 @@ package kube
 import (
 	"context"
 	"fmt"
+	"os"
 	"time"
 
 	kube "github.com/argoproj-labs/argocd-image-updater/registry-scanner/pkg/kube"
@@ -15,6 +16,7 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
 	_ "k8s.io/client-go/plugin/pkg/client/auth"
+	"k8s.io/client-go/tools/clientcmd"
 )
 
 type ImageUpdaterKubernetesClient struct {
@@ -22,11 +24,40 @@ type ImageUpdaterKubernetesClient struct {
 	KubeClient            *kube.KubernetesClient
 }
 
-func NewKubernetesClient(ctx context.Context, client kubernetes.Interface, applicationsClientset versioned.Interface, namespace string) *ImageUpdaterKubernetesClient {
+func NewKubernetesClient(ctx context.Context, kubeConfig, namespace string) (*ImageUpdaterKubernetesClient, error) {
+
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	loadingRules.DefaultClientConfig = &clientcmd.DefaultClientConfig
+	loadingRules.ExplicitPath = kubeConfig
+	overrides := clientcmd.ConfigOverrides{}
+	clientConfig := clientcmd.NewInteractiveDeferredLoadingClientConfig(loadingRules, &overrides, os.Stdin)
+
+	config, err := clientConfig.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	if namespace == "" {
+		namespace, _, err = clientConfig.Namespace()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	applicationsClientset, err := versioned.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
 	kc := &ImageUpdaterKubernetesClient{}
-	kc.KubeClient = kube.NewKubernetesClient(ctx, client, namespace)
+	kc.KubeClient = kube.NewKubernetesClient(ctx, clientset, namespace)
 	kc.ApplicationsClientset = applicationsClientset
-	return kc
+	return kc, nil
 }
 
 // CreateApplicationEvent creates a kubernetes event with a custom reason and message for an application.

--- a/pkg/kube/kubernetes_test.go
+++ b/pkg/kube/kubernetes_test.go
@@ -1,7 +1,6 @@
 package kube
 
 import (
-	"context"
 	"testing"
 
 	appv1alpha1 "github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
@@ -14,22 +13,6 @@ import (
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
-
-func Test_NewKubernetesClient(t *testing.T) {
-	t.Run("Get new K8s client for remote cluster instance", func(t *testing.T) {
-		client, err := registryKube.NewKubernetesClientFromConfig(context.TODO(), "", "../../test/testdata/kubernetes/config")
-		require.NoError(t, err)
-		assert.NotNil(t, client)
-		assert.Equal(t, "default", client.Namespace)
-	})
-
-	t.Run("Get new K8s client for remote cluster instance specified namespace", func(t *testing.T) {
-		client, err := registryKube.NewKubernetesClientFromConfig(context.TODO(), "argocd", "../../test/testdata/kubernetes/config")
-		require.NoError(t, err)
-		assert.NotNil(t, client)
-		assert.Equal(t, "argocd", client.Namespace)
-	})
-}
 
 func Test_GetDataFromSecrets(t *testing.T) {
 	t.Run("Get all data from dummy secret", func(t *testing.T) {

